Quick monster upgrade in towns

a1exsh opened the pull request:
If an army troop is located in a town that provides a possibility to upgrade
them, display a quick upgrade button on the troop sprite.  Clicking the button
invokes the upgrade troop dialog directly.

a1exsh commented:
To my taste, this changes touches a bit too many places, but I haven't found a more direct way to achieve the same effect so far.

a1exsh commented:
Screenshot:
![image](https://user-images.githubusercontent.com/489601/130357842-f3e56059-f5a6-4559-bcb5-e30fad69703e.png)

a1exsh commented on src/fheroes2/castle/buildinginfo.h:
```diff
@@ -72,7 +72,7 @@ class DwellingsBar : public Interface::ItemsBar<DwellingItem>
     void RedrawItem( DwellingItem &, const fheroes2::Rect &, fheroes2::Image & ) override;
 
-    bool ActionBarLeftMouseSingleClick( DwellingItem & dwelling ) override;
+    bool ActionBarLeftMouseSingleClick( const fheroes2::Point &, DwellingItem & dwelling, const fheroes2::Rect & ) override;
```
Not my fault: it was like that before me :-p

ihhub commented:
Hi @a1exsh , I like your great enthusiasm about new ideas but we usually make a discussion at least between members of our team before implementing new things. Please next time propose your idea first in discussion topic :)

@Branikolog , @oleg-derevenetz and @LeHerosInconnu, could you please check this change and express your opinion about it?

a1exsh commented on src/fheroes2/castle/buildinginfo.h:
```diff
@@ -72,7 +72,7 @@ class DwellingsBar : public Interface::ItemsBar<DwellingItem>
     void RedrawItem( DwellingItem &, const fheroes2::Rect &, fheroes2::Image & ) override;
 
-    bool ActionBarLeftMouseSingleClick( DwellingItem & dwelling ) override;
+    bool ActionBarLeftMouseSingleClick( const fheroes2::Point & /*unused*/, DwellingItem & dwelling, const fheroes2::Rect &  /*unused*/) override;
```
```suggestion
    bool ActionBarLeftMouseSingleClick( const fheroes2::Point & /*unused*/, DwellingItem & dwl, const fheroes2::Rect &  /*unused*/) override;
```

a1exsh commented:
@ihhub I'm just using it as an opportunity to study the code ;)  I'm going to play a locally modified version for my convenience anyway

oleg-derevenetz commented:
@ihhub @a1exsh Having a working prototype is much better than having just an idea. However, it may happen that the idea will not be accepted. The author should be ready for this :)

LeHerosInconnu commented:
Hello everyone,

> Hi @a1exsh , I like your great enthusiasm about new ideas but we usually make a discussion at least between members of our team before implementing new things. Please next time propose your idea first in discussion topic :)
> 
> @Branikolog , @oleg-derevenetz and @LeHerosInconnu, could you please check this change and express your opinion about it?

It is not yet possible to try the change.
But I completely agree with this improvement as it was already on my list of ideas not yet proposed. :D

Branikolog commented:
Hi all. 
The proposal to fast upgrade all units is not bad at all, but we firstly should focus on fixing existing bugs, than adding numerous new features. 
Also, this feature should be relevant to castle GUI, which is not a trivial without artists. 
We can leave this feature for some hot key in a castle view window, if it doesn't break anything.

LeHerosInconnu commented:
Hello @Branikolog,

> Hi all.
> The proposal to fast upgrade all units is not bad at all, but we firstly should focus on fixing existing bugs, than adding numerous new features.
> Also, this feature should be relevant to castle GUI, which is not a trivial without artists.
> We can leave this feature for some hot key in a castle view window, if it doesn't break anything.

This is not to upgrade all creatures at once, but to visually indicate which troops can be upgraded.
The player can then choose whether or not to upgrade them one by one.

Branikolog commented:
> This is not to upgrade all creatures at once, but to visually indicate which troops can be upgraded.
> The player can then choose whether or not to upgrade them one by one.

Oh, sorry for my slight misunderstanding...
Anyway, we can add thousands of such nice features, but without any professional artist, who would add such elements to GUI without ruining game style, the majority of them could be left unimplemented.

a1exsh commented:
Yes, what is implemented here is a small "up" arrow in the top-bottom corner of an upgradable troop.  Clicking on that arrow opens the upgrade dialog directly without going through army dialog first where one needs to click Upgrade button (it saves 2 clicks).

ihhub commented on src/fheroes2/dialog/dialog_armyinfo.cpp:
```diff
@@ -169,20 +185,9 @@ int Dialog::ArmyInfo( const Troop & troop, int flags, bool isReflected )
-                        result = Dialog::UPGRADE;
-                        break;
-                    }
+                if ( Dialog::YES == Dialog::TroopUpgrade( troop, UPGRADE_DISABLE & flags ) ) {
```
Could you please fix compilation warning as it caught on Visual Studio:
```
src\fheroes2\dialog\dialog_armyinfo.cpp(188): error C4800: 'int': forcing value to bool 'true' or 'false' (performance warning) [C:\projects\fheroes2\fheroes2-vs2015.vcxproj]
```
this prevents the creation of installer for our QA team to check how it works :)

a1exsh replied:
fixed

LeHerosInconnu commented:
Hello everyone,

So, I did some tests.
Everything seems to work fine.
Good job on the singular plural for text for the information about the function of the button in the bottom bar depending on the number of creatures and also on the message when the player doesn't have enough resources to upgrade a troop. :)

Some suggestions and possible improvements (can be done in another PR for some of them).

- display the graphic in the upper left corner to avoid the arrow to cover the sprite of the creature, in the examples with Dragons, Cavalry and Paladin.
![Quick upgrade button 001](https://user-images.githubusercontent.com/43583994/130655059-7d545b33-dfc6-4ea6-ab44-69cf1f554a2c.png)
![Quick upgrade button 002](https://user-images.githubusercontent.com/43583994/130655083-1640c2af-e0ce-4cb5-8eed-70ba6272d92a.png)

- in the TOWNS/CASTLES screen, also display the graphic in the upper left corner, it would be also better to use a smaller graphic in this screen.
![Quick upgrade button 003](https://user-images.githubusercontent.com/43583994/130655160-9e35f4bc-e4ec-499f-a2bd-46a197f6176b.png)
![Quick upgrade button 004](https://user-images.githubusercontent.com/43583994/130655180-35b6ca66-ccf2-4fc9-b22c-d341a144b790.png)

- for the Green Dragon, it should be possible either to add a second graphic below the first one to allow the Green Dragon to be upgraded directly to Black Dragon or to add at least one button that allows this in the validation window.
![Quick upgrade button 005](https://user-images.githubusercontent.com/43583994/130655273-0d98e174-16c7-473e-8b9e-75b3b5a41316.png)
![Quick upgrade button 006](https://user-images.githubusercontent.com/43583994/130655297-2cf1cf12-bcb6-4a6e-8479-02f64e0934c2.png)

- a color code for the graphic could also be used to visually indicate to the player if he can upgrade the troop directly (yellow arrow), if he can upgrade the troop provided he exchanges resources (orange arrow), if he cannot upgrade the troop even by exchanging resources (red arrow).
![Quick upgrade button 007](https://user-images.githubusercontent.com/43583994/130655339-1654480c-a593-4da2-be89-30ce82660b30.png)

a1exsh commented:
@LeHerosInconnu thanks for having a look: you raise a number of good points!

It occurs to me now that all monsters in Heroes II are looking to the right, so probably it makes sense to move the quick upgrade button to the top left corner.
Now I see where the mini army bar is used: in kingdom overview.  Haven't found this in the code. :sweat_smile: 

Will check the green -> black dragon interaction.  I also like the idea to give a (color) hint when upgrade is not possible due to lack of resources.

LeHerosInconnu commented:
Hello @a1exsh,

> Now I see where the mini army bar is used: in kingdom overview. Haven't found this in the code. 

You made me think that a mini creature army bar is also displayed in the hero meeting screen.

In the following example, Rebecca is located in a castle where the Irons Golems can be upgraded.
![Quick upgrade button 008](https://user-images.githubusercontent.com/43583994/130683652-5000aad6-80e7-45d7-a1eb-3754927bb3bb.png)

But during the meeting with another hero (Rebecca stays in the castle), the graphic for upgrading the Iron Golems is not displayed, while it is possible to upgrade the Iron Golems with the default method.
The upgrade graphics should also be displayed in the hero meeting screen.
![Quick upgrade button 009](https://user-images.githubusercontent.com/43583994/130683852-ba52c326-e554-4d41-ade6-10564f59bf6a.png)
![Quick upgrade button 010](https://user-images.githubusercontent.com/43583994/130683873-64d42fdd-3715-4938-a278-93fc497f2ce4.png)

oleg-derevenetz commented on src/fheroes2/army/army_bar.cpp:
```diff
@@ -185,6 +185,44 @@ bool ArmyBar::isValid() const
 
+const fheroes2::Sprite ArmyBar::GetUpgradeButton()
+{
+    static const fheroes2::Sprite upButton = fheroes2::AGG::GetICN( ICN::RECRUIT, 0 );
```
Oh, these "suggested changes"... Sometimes they are completely broken. I'll have to take a look sometime.

ihhub replied:
Wow!

oleg-derevenetz replied:
Hi @ihhub I (hopefully) fixed such issues with suggested changes, corresponding PR is merged to the upstream, you may want to sync [fheroes2/clang-tidy-pr-comments](https://github.com/fheroes2/clang-tidy-pr-comments) with the upstream.

a1exsh commented:
@LeHerosInconnu I've moved the button to the top-left corner, as suggested, and also made it smaller for the mini-sprite rendering.  The icon is now also rendering when in the hero meeting dialog. :)

It looks like the icon now fits nicely in all cases I have observed (and it's clickable):
![image](https://user-images.githubusercontent.com/489601/130743135-264c0176-d29c-467e-b835-749263ce0302.png)
![image](https://user-images.githubusercontent.com/489601/130743319-3dea86ec-7c8c-4b26-9238-5697b0655028.png)
![image](https://user-images.githubusercontent.com/489601/130743400-bd7d9cf9-dac4-4af4-86ff-4e3edc83b9cc.png)

For now I'd like not to improve it further, as to add skip-upgrade for green->black dragon or to add color code for lacking resources on upgrade.  The existing upgrade button doesn't have these features either, and it can be done separately later.

I am noticing though that a _guardian_ hero is not allowed to upgrade the troops in the army, even with the existing dialog.  Do you know if that is by design or should be fixed (separately)?

Are guardians allowed to leave the castle at all?

oleg-derevenetz commented on src/fheroes2/army/army_bar.cpp:
```diff
@@ -185,6 +185,44 @@ bool ArmyBar::isValid() const
+{
+    static const fheroes2::Sprite upButton = fheroes2::AGG::GetICN( ICN::RECRUIT, 0 );
+    if ( use_mini_sprite ) {
+        static fheroes2::Sprite miniUpButton( upButton.width() / 2, upButton.height() / 2 );
```
May be use lambda here for initialization of `miniUpButton` instead of this ugly construction with `static bool` flag? In this case you can even make it `const`.

a1exsh replied:
I wouldn't call it ugly, though I'm open to suggestions how to make it better. :)

oleg-derevenetz replied:
Something like this (out of my head, no syntax validation performed):

```
static const fheroes2::Sprite miniUpButton = [&upButton]() {
    fheroes2::Sprite result(upButton.width() / 2, upButton.height() / 2);

    fheroes2::Resize( upButton, 0, 0, upButton.width(), upButton.height(), result, 0, 0, result.width(), result.height(), true );

    return result;
}();
```

a1exsh replied:
@oleg-derevenetz that worked, thanks a lot! :)

ihhub commented on src/fheroes2/heroes/heroes_meeting.cpp:
```diff
@@ -140,6 +140,12 @@ class MeetingArmyBar : public ArmyBar
 
+        if ( CanUpgradeNow( troop ) ) {
+            const fheroes2::Rect upButtonPos = GetUpgradeButtonPos( roi );
+            const fheroes2::Sprite upgradeButton = GetUpgradeButton();
```
Please use const reference to avoid object copy.

ihhub commented on src/fheroes2/castle/castle_dialog.cpp:
```diff
@@ -682,6 +682,12 @@ int Castle::OpenDialog( bool readonly )
                     RedrawIcons( *this, heroes, cur_pt );
+
+                // may render new quick upgrade possibilities
+                selectArmy1.Redraw();
+                if ( selectArmy2.isValid() && alphaHero >= 255 )
+                    selectArmy2.Redraw();
```
What's about hero hiring? Do we need to do the same?

a1exsh replied:
No, a hero just hired gets everything rendered properly.  It's only when you upgrade a building the new upgrades may need to be reflected.

ihhub commented on src/fheroes2/dialog/dialog.h:
```diff
@@ -107,6 +107,7 @@ namespace Dialog
     Troop RecruitMonster( const Monster &, u32 available, bool );
     void DwellingInfo( const Monster &, u32 available );
     bool SetGuardian( Heroes &, Troop &, CapturedObject &, bool readonly );
+    int TroopUpgrade( const Troop & troop, bool cannotAfford );
```
`cannotAfford` could be `const` as well.

ihhub commented on src/fheroes2/dialog/dialog_armyinfo.cpp:
```diff
@@ -71,6 +71,22 @@ void DrawMonsterInfo( const fheroes2::Point & dst, const Troop & troop );
 
+int Dialog::TroopUpgrade( const Troop & troop, bool cannotAfford )
+{
+    std::string msg;
+    int flags;
+
+    if ( cannotAfford ) {
+        msg = _( "You can't afford to upgrade your troops!" );
+        flags = Dialog::OK;
+    }
+    else {
+        msg = _( "Your troops can be upgraded, but it will cost you dearly. Do you wish to upgrade them?" );
+        flags = Dialog::YES | Dialog::NO;
+    }
+    return Dialog::ResourceInfo( "", msg, troop.GetUpgradeCost(), flags );
```
We could simplify this function to be:
```cpp
if ( cannotAfford ) {
    return Dialog::ResourceInfo( "", _( "You can't afford to upgrade your troops!" ), troop.GetUpgradeCost(), Dialog::OK );
}

return Dialog::ResourceInfo( "", _( "Your troops can be upgraded, but it will cost you dearly. Do you wish to upgrade them?" ), troop.GetUpgradeCost(), Dialog::YES | Dialog::NO);
```
so we won't need local variables here.

ihhub reviewed (changes requested):
Hi @a1exsh , I left just few comments here. Could you please take a look?

ihhub commented:
@a1exsh ,
> I am noticing though that a guardian hero is not allowed to upgrade the troops in the army, even with the existing dialog. Do you know if that is by design or should be fixed (separately)?
 >Are guardians allowed to leave the castle at all?

This is an experimental option which haven't tested fully. Guardians should be allowed to leave the castle.

LeHerosInconnu commented:
Hello @a1exsh,

> @LeHerosInconnu I've moved the button to the top-left corner, as suggested, and also made it smaller for the mini-sprite rendering. The icon is now also rendering when in the hero meeting dialog. :)
> 
> It looks like the icon now fits nicely in all cases I have observed (and it's clickable):

Works well, just a few remarks. :)
The position of the small upgrade button for screens with miniature sprites should be shifted one pixel to the left to have the same two pixels space from the top of the frame and from the left of the frame.
(This would also allow the button to be a little further away from the Paladin sprite in the example.)

Also, if it were possible to make the click area a little larger (in all screens) by extending it around the button in the upper left corner (red rectangles in the examples) and not just the button sprite, it would make it more convenient to use (no need to concentrate on clicking precisely just on the button area).
![Quick upgrade area 100](https://user-images.githubusercontent.com/43583994/130818469-6373e750-7aac-40fc-bd3a-61e5e0acad1e.png)
![Quick upgrade area 101](https://user-images.githubusercontent.com/43583994/130818488-fee64431-0304-4af4-b116-e0f6e1cf4ce3.png)
![Quick upgrade area 102](https://user-images.githubusercontent.com/43583994/130818501-92fd9e89-8f34-424c-9f41-7a651f1e84b6.png)

> For now I'd like not to improve it further, as to add skip-upgrade for green->black dragon or to add color code for lacking resources on upgrade. The existing upgrade button doesn't have these features either, and it can be done separately later.

No problem, this can be done later. :)

> I am noticing though that a _guardian_ hero is not allowed to upgrade the troops in the army, even with the existing dialog. Do you know if that is by design or should be fixed (separately)?

Good catch, it's a bug.
I think it would be better to fix it in another PR.
Everything should work for the guardian hero in the same way as for the Captain and a hero visiting the castle.
![Quick upgrade 105](https://user-images.githubusercontent.com/43583994/130819780-2caabfbd-e984-4d28-9144-55d6be04cbb0.png)
![Quick upgrade 104](https://user-images.githubusercontent.com/43583994/130819753-9e93a699-8fa0-4d79-a879-598bec005632.png)
![Quick upgrade 103](https://user-images.githubusercontent.com/43583994/130819735-21b0223f-66f3-47fd-8e73-24dc43a30dde.png)

> Are guardians allowed to leave the castle at all?

Yes, the hero guardian can leave the garrison (Captain's location) by clicking on the slot below it at any time.
The guardian hero is counted as one of the eight heroes that a player can recruit at the same time.
![Quick upgrade 106](https://user-images.githubusercontent.com/43583994/130818982-c0481eaa-02ab-472a-96d4-6f9da43a847b.png)
![Quick upgrade 107](https://user-images.githubusercontent.com/43583994/130819011-a336d26d-e391-4f20-9de5-f3809e3bc76c.png)

a1exsh commented:
Thanks for the review, @ihhub and @LeHerosInconnu !

I've addressed the code comments and also made the clickable area of the button a bit bigger, as suggested.  I will also try to see if we can cast a shadow under the button to make it stand out a little more, like on the actual "recruit troop" dialog.

oleg-derevenetz commented:
@a1exsh if you sync this PR with `master` branch, quick upgrades should work with guardian heroes too.

ihhub commented on src/fheroes2/army/army_bar.cpp:
```diff
@@ -190,6 +192,53 @@ bool ArmyBar::isValid() const
     return _army != nullptr;
 }
 
+const fheroes2::Sprite & ArmyBar::GetUpgradeButton() const
```
Also this method should be private as it's not used anywhere outside the class.

a1exsh replied:
I guess there was no memory duplication either way, because of `static` storage used inside the function.  Moved it to private namespace, I hope that also works with you? :)

a1exsh commented:
> I think we still need to make it as a button. I'll check in a while how to render such properly.

@ihhub in the OG editor there is a clickable button (it has a pressed state image as well) for drop down menus:
![image](https://user-images.githubusercontent.com/489601/132171943-5694373f-0b86-47bc-be14-8e1535081d88.png)

I guess it should be part of the game resources, so we could use it here if we mirrored it vertically.

Do you by chance know the icon name for this one?

ihhub commented:
> 
> 
> > I think we still need to make it as a button. I'll check in a while how to render such properly.
> 
> @ihhub in the OG editor there is a clickable button (it has a pressed state image as well) for drop down menus:
> ![image](https://user-images.githubusercontent.com/489601/132171943-5694373f-0b86-47bc-be14-8e1535081d88.png)
> 
> I guess it should be part of the game resources, so we could use it here if we mirrored it vertically.
> 
> Do you by chance know the icon name for this one?

`ICN::LISTBOX` is what you're looking for.

a1exsh commented:
Updated the button icon, thanks for the hint, @ihhub ! :-)
![image](https://user-images.githubusercontent.com/489601/132465143-b5a1462d-8c10-4498-89e8-27bfb4138900.png)

In order to render pressed and released state I had to further extend `ActionBarLeftMouseHold` with `cursor` and item position parameters, though this approach doesn't seem to scale really well.

I see that we have a special class for the buttons here: https://github.com/ihhub/fheroes2/blob/master/src/fheroes2/gui/ui_button.h#L99
Using it, however, requires fixed position on the screen.  For us to use it in the `ArmyBar`, we would need to keep a mapping between the troops and the buttons.  That would also remove the need to pass item position in the Click/Hold event handlers.

I will have a look later.

GeorgeK1ng commented:
Wow! I really like these improvements!
